File: Haesoo/Week15/cpp/BOJ_ACM_Craft.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <map>
using namespace std;
#define fastio ios::sync_with_stdio(0), cin.tie(0)

int main() {
	int T;
	cin >> T;
	for (int j = 0; j < T; j++) {
		int N, K, W;
		cin >> N >> K;
		//vector <int> depth(N);
		int depth[1000] = { 0, };
		vector <vector <int>> relation(N);
		//map <int, int> cost;
		//map <int, int> result;
		int cost[1000] = { 0, };
		int result[1000] = { 0, };
		for (int i = 0; i < N; i++) {
			cin >> cost[i];
			result[i] = cost[i];
		}
		for (int i = 0; i < K; i++) {
			int prev, next;
			cin >> prev >> next;
			depth[next - 1]++;
			relation[prev - 1].push_back(next - 1);
		}
		cin >> W;

		queue <int> q;
		for (int i = 0; i < N; i++) {
			if (i == W - 1) continue;
			if (depth[i] == 0) q.push(i);
		}
		while (!q.empty()) {
			int curNode = q.front();
			q.pop();
			for (int i = 0; i < relation[curNode].size(); i++) {
				int next = relation[curNode][i];
				result[next] = max(result[next], result[curNode] + cost[next]);
				if (--depth[next] == 0) q.push(next);
			}
		}
		cout << result[W - 1] << endl;
	}
}

File: Haesoo/Week15/BOJ_ACM-Craft.java
import java.util.*;
import java.io.*;

public class BOJ1005 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int T = Integer.parseInt(br.readLine());
        for (int i = 0; i < T; i++) {
            st = new StringTokenizer(br.readLine());
            int N = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());
            int[] depth = new int[1000];
            List<Integer>[] relations = new List[N];
            int[] cost = new int[1000];
            int[] result = new int[1000];
            Queue<Integer> q = new ArrayDeque<>();
            for (int j = 0; j < N; j++) {
                relations[j] = new ArrayList<>();
            }
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                cost[j] = Integer.parseInt(st.nextToken());
                result[j] = cost[j];
            }
            for (int j = 0; j < K; j++) {
                st = new StringTokenizer(br.readLine());
                int prev = Integer.parseInt(st.nextToken()) - 1;
                int next = Integer.parseInt(st.nextToken()) - 1;
                depth[next]++;
                relations[prev].add(next);
            }
            int W = Integer.parseInt(br.readLine()) - 1;
            for (int j = 0; j < N; j++) {
                if (j == W) continue;
                if (depth[j] == 0) q.add(j);
            }
            while (!q.isEmpty()) {
                int curNode = q.peek();
                q.poll();
                for (int j = 0; j < relations[curNode].size(); j++) {
                    int next = relations[curNode].get(j);
                    result[next] = Math.max(result[next], result[curNode] + cost[next]);
                    if (--depth[next] == 0) q.add(next);
                }
            }
            System.out.println(result[W]);
        }
    }
}
